### PR TITLE
Support CUDA tensors in ProcessGroupGloo 

### PR DESCRIPTION
**Pull request by pietern**
Depends on #7628 being merged.

This also includes everything in #7628 because of this dependency. Once that is merged this will be rebased.

### CODE REVIEW

**Comment by pietern** [comment]
I copied the AutoGPU struct from torch/csrcs/utils and ideally source it from some place instead of keeping a copy in this tree. It can also be AutoDevice instead of AutoGPU, depending on tensor type.

File: torch/lib/c10d/CMakeLists.txt
Comment (apaszke) [inline torch/lib/c10d/CMakeLists.txt]: CUDA and gloo are required only for now, and we'll make them optional in the future, right?

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (apaszke) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: Why a `unique_ptr` and an extra allocation? It has a default constructor that doesn't do anything, better to use that.

File: torch/lib/c10d/CMakeLists.txt
Comment (pietern) [reply torch/lib/c10d/CMakeLists.txt]: Yep, that's the plan. Don't yet know how, but I want to have backend specific libraries like ATen.

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: This is now fixed.

File: torch/lib/c10d/ProcessGroup.hpp
Comment (apaszke) [inline torch/lib/c10d/ProcessGroup.hpp]: Can you add some more docs for this function? Does it synchronize CUDA/CPU/both? Is there a description for the concurrency model of ProcessGroups somewhere? Can you refer to it from here? Would I ever want to call this manually instead of using `wait()`?

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (apaszke) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: This function only inserts public -> private synchronization, so the comment is misleading.

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (apaszke) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: This means gloo will only be possible to compile when someone has CUDA. Do we really want to enforce that?

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (apaszke) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: Shouldn't this be part of the `run` body? Why do we only wait for public streams only once?

File: torch/lib/c10d/ProcessGroupGloo.hpp
Comment (apaszke) [inline torch/lib/c10d/ProcessGroupGloo.hpp]: Are those events/streams E1/S1 or E2/S2 from the description above?

File: torch/lib/c10d/ProcessGroupGloo.hpp
Comment (apaszke) [inline torch/lib/c10d/ProcessGroupGloo.hpp]: Similar question like above. Would be good to have a reference to the note at the top.

**Review by apaszke — COMMENTED** [review]
@pietern @teng-li you might want to add yourself as CODEOWNERS of the c10d dir, because it will make your own approves unblock the merge.

File: torch/lib/c10d/CMakeLists.txt
Comment (teng-li) [inline torch/lib/c10d/CMakeLists.txt]: Do we want to fail the build if CUDA cannot be found? I am thinking of making it optional

File: torch/lib/c10d/ProcessGroup.hpp
Comment (teng-li) [reply torch/lib/c10d/ProcessGroup.hpp]: Yeah, i was gonna ask the same

Comment (pietern) [reply torch/lib/c10d/ProcessGroup.hpp]: Added:

```
+    // For CUDA tensors, it inserts stream synchronization such that
+    // the streams of the caller wait for completion of the
+    // asynchronous operations on the destination tensors.
+    //
+    // For CPU tensors, it is currently a nop.
+    //
+    // This function should only be used if the caller polls for
+    // completion through the `isCompleted` function, it has returned
+    // true, and the `isSuccess` function also has returned true.
+    //
```

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: Thanks for catching this. This was my original intent, but proved to be impossible, so changed it later.

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: For now, I'm fine with this.

I want to get a first rev out and making it depend on CUDA is reasonable IMO.

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: We need to capture the state of the caller's streams at the call site. Everything in the run body is run by a different thread, where we have a different thread local streams.

File: torch/lib/c10d/ProcessGroupGloo.hpp
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.hpp]: Yes, updating the comment to be clearer.

File: torch/lib/c10d/ProcessGroupGloo.hpp
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.hpp]: Added comment clarifying the use of local events here.

File: torch/lib/c10d/CUDAUtils.hpp
Comment (pietern) [reply torch/lib/c10d/CUDAUtils.hpp]: It is const w.r.t. the CUDAEvent instance, but returns a non-const pointer.

File: torch/lib/c10d/CUDAUtils.hpp
Comment (pietern) [reply torch/lib/c10d/CUDAUtils.hpp]: Good point, changing this.

File: torch/lib/c10d/CMakeLists.txt
Comment (pietern) [reply torch/lib/c10d/CMakeLists.txt]: Right now I think it is acceptable, so we can get a first working version out the door.

**Comment by pietern** [comment]
@apaszke @teng-li The latest 2 commits should address your comments, can you check?

Looking into the CODEOWNERS thing now.

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (teng-li) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: It would be nice to support GPU direct when creating the algorithm and when IB is available.  Can you create a TODO to automatically enable GPU direct when the device supports it?

**Review by teng-li — APPROVED** [review]
Looks great!

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (teng-li) [inline torch/lib/c10d/ProcessGroupGloo.cpp]: maybe you can also do
```
deviceGuard.setDevice(device);
auto publicStream = THCState_getCurrentStream(thcState):
```

Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: I prefer the current version since it does 1 less cudaGetDevice call (even though it is cheap)

File: torch/lib/c10d/ProcessGroupGloo.cpp
Comment (pietern) [reply torch/lib/c10d/ProcessGroupGloo.cpp]: Added #8032

**Pull request merged**
